### PR TITLE
Refactor/move path to hdi

### LINKED ISSUE #2397: [Enhancement] Include Path, Component in hdi
It would be nice to have Path and Component imported into hdi. If you use a Path in coordinator you're most likely going to need to do some kind of validation of that in integrity, so makes sense to include the types in both.

### LINKED ISSUE #2519: Request to add some path functionality into HDI
**Describe the bug**

This was a question to Guillem and he asked me to file it with Holochain team.
In my app I need to create a link based on the content of the target entry. For example: 

```
let badge = Badge {
    name: "Alastair",
    location: "London",
    ...
  }; 

let action_hash = create_entry(EntryTypes::Badge(badge.clone()))?;
let anchor_str = format!("{}{}", badge.name, badge.location);;
let anchor = Path::from(&anchor_string).path_entry_hash()?
create_link(
    anchor, 
    action_hash,        
    None,
    None,
)?;
```
In validation I would like to check that the link anchor is in fact the correct one for the badge. So something like...

```
let link_base = // get link base

    let link_target_badge = badge_record.entry().to_app_option::<Badge>().unwrap().unwrap();

    let base_path_string = format!("{}{}", link_target_badge.name, link_target_badge.location);
    let base_path = Path::from(base_path_string).path_entry_hash()?;

if base_path == link_base {
  Ok(ValidationCallbackResult::Valid)
}
```

But `let base_path = Path::from(base_path_string).path_entry_hash()?;` doesn't work because Path doesn't exist in the HDI. 
**Expected behavior**

**System information:**
holochain (holochain 0.1.5): db5b8b27da3bf296958c3bf54ac3950dc60a39c8
lair-keystore (lair_keystore 0.2.4): 43be404da0fd9d57bf4429c44def405bd6490f61
hc-launch (holochain_cli_launch 0.0.12): 1ad188a43900c139e52df10a21e3722f41dfb967
hc-scaffold (holochain_scaffolding_cli 0.1.8): 85997cbc4c92f0fea87447d7c7daed1245b47700

{"git_info":null,"cargo_pkg_version":"0.1.5","hdk_version_req":"0.1.3","hdi_version_req":"0.2.3","timestamp":"2023-06-12T10:14:14.311672Z","hostname":"Alastairs-MacBook-Pro.local","host":"x86_64-apple-darwin","target":"x86_64-apple-darwin","rustc_version":"rustc 1.66.1 (90743e729 2023-01-10)","rustflags":"","profile":"release"}

**Steps to reproduce**
N/A

### PR DESCRIPTION
**Pull request by mattyg**
### Summary
Conclusion of #2613 -- this is a breaking change that can't be backported I believe. 

Moves type defs of Anchor, Path, Component into hdi. Creates some new traits in hdk to hold the functionality that required hdk.

Resolves #2519 
Resolves #2397

### TODO:
- [x] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: crates/hdi/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
 
+- Add types for hash paths (migrated from hdk crate) and include in prelude: Anchor, Path, Component, TypedPath
```
Comment (jost-s) [inline crates/hdi/CHANGELOG.md:10]:
```suggestion
- Migrate types for hash paths from hdk crate and include in prelude: Anchor, Path, Component, TypedPath
```

**Review by jost-s — APPROVED** [review]
LGTM

**Comment by ThetaSinner** [comment]
This looks great! 

There is still a bit of work to update all the references to the moved types which will be the demonstration that the change is complete. Happy to help with that part if that would be useful, it's hopefully just a lot of fiddly replacements :)

**Comment by mattyg** [comment]
> There is still a bit of work to update all the references to the moved types which will be the demonstration that the change is complete. Happy to help with that part if that would be useful, it's hopefully just a lot of fiddly replacements :)

Okay added the types to hdk prelude and fixed the one other place where it was used directly. Should be good now.

**Pull request merged**